src/jit/hwintrinsicxarch.cpp

```diff
@@ -1030,6 +1032,119 @@ GenTree* Compiler::impSSE2Intrinsic(NamedIntrinsic        intrinsic,
             break;
         }
 
+        case NI_SSE2_SetAllVector128:
+        {
+            assert(sig->numArgs == 1);
+            baseType = getBaseTypeOfSIMDType(sig->retTypeSigClass);
+            assert((baseType >= TYP_BYTE) && (baseType <= TYP_DOUBLE));
+            assert(baseType != TYP_FLOAT);
+
+            op1          = impPopStack().val;
+            GenTree* src = nullptr;
+
+            switch (baseType)
+            {
+                case TYP_DOUBLE:
+                {
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_SetScalarVector128, baseType, simdSize);
+                    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, src, gtNewIconNode(0b01000100), NI_SSE2_Shuffle,
+                                                       TYP_INT, simdSize);
+                    break;
+                }
+
+                case TYP_LONG:
+                case TYP_ULONG:
+                {
+                    var_types shuffleBaseType = TYP_UNKNOWN;
+                    if (baseType == TYP_LONG)
+                    {
+                        intrinsic       = NI_SSE2_ConvertScalarToVector128Int64;
+                        shuffleBaseType = TYP_INT;
+                    }
+                    else
+                    {
+                        intrinsic       = NI_SSE2_ConvertScalarToVector128UInt64;
+                        shuffleBaseType = TYP_UINT;
+                    }
+#ifdef _TARGET_X86_
+                    src     = gtNewSimdHWIntrinsicNode(TYP_SIMD16, op1, NI_SSE2_SetScalarVector128, TYP_DOUBLE, simdSize);
```

Comment:
Why? `movq` is available on both 32-bit and 64-bit (it is one of the 64-bit instructions that is).

![image](https://user-images.githubusercontent.com/10487869/37054207-1202b674-2133-11e8-8288-e8221062d3b6.png)

Reply:
You may have been looking at `movd`, which has a `movq` equivalent encoding, which is 64-bit only.

Reply:
This code is problemit as it does not work as expected on x86 for practically all possible options I have tested so far including `movq` instruction.

For `TYP_LONG` used for both `long` and `ulong` on x86 code always hits assertion:

```
e:\src\ms\dotnet\coreclr\bin\tests\Windows_NT.x64.Checked\TestWrappers\JIT.HardwareIntrinsics\JIT.HardwareIntrinsics.XUnitWrapper.cs(14037): error : JIT_HardwareIntrinsics._X86_Sse2_SetAllVector128_r_
SetAllVector128_r_._X86_Sse2_SetAllVector128_r_SetAllVector128_r_cmd [FAIL] [e:\src\ms\dotnet\coreclr\tests\runtest.proj]

        Assert failure(PID 10964 [0x00002ad4], Thread: 19116 [0x4aac]): Assertion failed 'tree->IsUnusedValue()' in 'IntelHardwareIntrinsicTest.Program:Main(ref):int' (IL size 1306)

            File: e:\src\ms\dotnet\coreclr\src\jit\lsraxarch.cpp Line: 132
            Image: e:\src\ms\dotnet\coreclr\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\CoreRun.exe
```

I have tried all relevant loading instructions, copying the operand (it was used so perhaps another copy should be available) but it always hits the same assert - except when I do change `op.gtType` type what causes other asserts when jit checks for type usage consistency and is not acceptable solution anyway.

Reply:
Well, there aren't any 64 bit integer registers on x86 so it's pretty obvious that this isn't going to work...

Reply:
@fiigii @CarolEidt @tannergooding @mikedn 

Any ideas what could be causing that assert?

Reply:
The assert occurs because a `GT_LONG` is not contained, but is also not a "top-level" (unused) node. I suspect that the handling of the parent node is failing to handle the `GT_LONG` case where it must be contained (and the individual halves treated as separate operands, in both the source count in LSRA as well as in codegen.

Reply:
@CarolEidt 

Thanks for explanation. I understand it is missing something similar to:

https://github.com/dotnet/coreclr/blob/28a1d145c82c6c75325dce58d9dcd53551320683/src/jit/lowerxarch.cpp#L2207

Reply:
Yes, similar, though of course the handling of the lo and hi operands will obviously be different for this case.